thorn/src/kernel/mailbox.c

```diff
@@ -1,52 +1,56 @@
 #include "kernel/mailbox.h"
 
-bool mailbox_request (volatile unsigned int * data_ptr, channel_t channel) {
-    volatile unsigned int * buffer = data_ptr;
-
+bool mailbox_request (volatile unsigned int * buffer, channel_t channel) {
     if (DUMP_BUFFER) {
         LOG ("Dumping message buffer before sending:");
         int_dump ((unsigned int *) buffer);
         hex_dump ((byte_t *) buffer);
     }
 
-    // 28-bit address (MSB) and 4-bit value (LSB)
-    unsigned int outgoing = ((unsigned int) ((long) data_ptr) & ~0xF) | (channel & 0xF);
-
-    // Wait until we can write
-    while (get32 (MBOX0 + MBOX_STATUS) & MBOX_WRITE_FULL)
-        ;
+    mailbox_message_t message = {buffer, channel};
 
-    // Write the address of our buffer to the mailbox with the channel appended
-    put32 (MBOX0 + MBOX_WRITE, outgoing);
-
-    unsigned int incoming;
-    while (1) {
-        // Is there a reply?
-        while (get32 (MBOX0 + MBOX_STATUS) & MBOX_READ_EMPTY)
-            ;
-
-        incoming = get32 (MBOX0 + MBOX_READ);
-
-        // Is it a reply to our message?
-        if (outgoing == incoming) {
-            if (buffer[1] != MBOX_SUCCESS) {
-                if (DUMP_BUFFER) {
-                    LOG ("Dumping failed message buffer:");
-                    int_dump ((unsigned int *) buffer);
-                    hex_dump ((byte_t *) buffer);
-                }
-
-                return false;
-            }
-            break;
-        }
-    }
+    mailbox_write (&message);
+    mailbox_read (&message);
 
     if (DUMP_BUFFER) {
         LOG ("Dumping successful message buffer:");
         int_dump ((unsigned int *) buffer);
         hex_dump ((byte_t *) buffer);
     }
 
-    return true;
+    if (buffer[1] == PROPERTY_RESPONSE_SUCCESS) {
+        return true;
+    } else {
+        return false;
+    }
```

Comment:
```suggestion
    return buffer[1] == PROPERTY_RESPONSE_SUCCESS;
```

Reply:
non branching version:
```suggestion
    return !(buffer[1] ^ PROPERTY_RESPONSE_SUCCESS);
```